compiler/next/include/chpl/types/TypeTag.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_TYPE_TYPETAG_H
+#define CHPL_TYPE_TYPETAG_H
+
+namespace chpl {
+namespace types {
+namespace typetags {
+
+
+/**
+
+  This enum is used to identify which ASTType subclass a node is.
+ */
+enum TypeTag {
+  // define the enum for all of the non-virtual Type nodes
+  // using macros and TypeClassesList.h
+  /// \cond DO_NOT_DOCUMENT
+  #define TYPE_NODE(NAME) NAME ,
+  #define TYPE_BEGIN_SUBCLASSES(NAME) START_##NAME ,
+  #define TYPE_END_SUBCLASSES(NAME) END_##NAME ,
+  /// \endcond
+  // Apply the above macros to TypeClassesList.h
+  #include "chpl/types/TypeClassesList.h"
+  // clear the macros
+  #undef TYPE_NODE
+  #undef TYPE_BEGIN_SUBCLASSES
+  #undef TYPE_END_SUBCLASSES
+  NUM_TYPE_TAGS
+};
```

Comment:
I'm a little concerned about this.  I've found `lldb` has trouble with enum names in the old version of the compiler - here's an example:
```
(lldb) p call->tryTag
(TryTag) $1 = TRY_TAG_NONE
(lldb) p call->tryTag == TRY_TAG_NONE
error: <user expression 3>:1:17: use of undeclared identifier 'TRY_TAG_NONE'
call->tryTag == TRY_TAG_NONE
                ^
```
In the above output, lldb knows how to print the value of the tryTag field, but it doesn't know how to use that same value from the command prompt.  In order to actually ensure that the value matches, we have to use the number that corresponds to `TRY_TAG_NONE` (which in this case is 0):
```
(lldb) p call->tryTag == 0
(bool) $2 = true
```
By having the enum order generated, it's more difficult to determine what number should be used.

It could very well be that there's some set up we've missed for LLDB that enables us to use enum constant names like I want to, or that there's a way to print the number that all the constants of an enum correspond to.  But if that's not the case, I'd like us to have a way of determining the numbers

Reply:
It works in GDB (with the asttags enum; e.g. `print asttags::Module`). I think lldb has some bugs here. I don't see how storing the numbers in the source code is possible while generating these based on ASTClassesList / TypeClassesList and I'd rather generate them than have that work in LLDB.

There are several solutions I can think of that we could do at some point (I don't think we should do any of these in this PR):
 * have a test program that uses this macro facility to print out the name and number of each enum value
 * have a function available for calling from the debugger that goes from a string to the enum value

Reply:
Yeah, I think that's fair to not resolve in this PR.  Just as long as it's something we're keeping track of so that we can get it to work when the new compiler is "ready to go" or whatever